### PR TITLE
IOS-1859 :: Feature Internal :: Add RW Testing Outcome in Harmony Swift

### PR DESCRIPTION
**Pull request by vilanovi**
**Merge / Pull request information**

* Asana task link: https://app.asana.com/0/1109863238977521/1202462981511859/f

Included in the PR:
- Adding number random generator in Harmony.
- Adding DataSoruce and Repository Spy classes in HarmonyTesting.
- Adding Mock ObjectValidation class in HarmonyTesting.
- Adding unit tests for cache and in-memory data source.
- Created new QueryNotSupported error
- Edited InMemoryDataSource to return QueryNotSupported error instead of fatal error

The goal of PR is to gather all the outcomes from the RW testing session and include them in the repo. If you guys have more tests to add or want to replace mine for others, you're welcome to do it (I'm not an expert in testing).

### CODE REVIEW

**Review by franmontiel — APPROVED** [review]

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+class InMemoryDataSourceTests: XCTestCase {
+    
+    private func provideEmptyDataSource() -> InMemoryDataSource<Int> {
+        return InMemoryDataSource<Int>()
+    }
+    
+    func test_put_value() throws {
+        // Given
+        let dataSoruce = provideEmptyDataSource()
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: `"id"` should be a random string.

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        let value = 42
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: `"value"` should be a random value.

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        // Then
+        let result = try dataSoruce.get(query).result.get()
+        expect(result).to(equal(value))
+    }
+    
+    func test_delete_value() throws {
+        // Given
+        let dataSoruce = provideEmptyDataSource()
+        let query = IdQuery("id")
+        let value = 42
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: same as above

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            XCTAssert(false)
+        } catch {
+            expect(error).to(beAnInstanceOf(CoreError.NotFound.self))
+        }
+    }
+    
+    func test_get_value_not_found() throws {
+        // Given
+        let dataSoruce = provideEmptyDataSource()
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: same as above

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        
+        // When
+        do {
+            _ = try dataSoruce.get(query).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.NotFound.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: This code could be improved with the following style:

```swift
        var error: Error?
        _ = try dataSoruce.get(query).result.get(error: &error)
        
        let expectedError = CoreError. NotFound("XXXX")
        expect(error).to(matchError(expectedError))
        
        
        // or
        expect { try dataSoruce.get(query).result.get() }.to(throwError(errorType: CoreError. NotFound.self))
```

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        let query = VoidQuery()
+        
+        // When
+        do {
+            _ = try dataSoruce.get(query).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.QueryNotSupported.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: same as above

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        let value = Int.random()
+        
+        // When
+        do {
+            _ = try dataSoruce.put(value, in: query).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.QueryNotSupported.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: same as above

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: empty break line

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        let query = VoidQuery()
+        
+        // When
+        do {
+            _ = try dataSoruce.delete(query).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.QueryNotSupported.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/InMemoryDataSourceTests.swift]: same as above

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+        }
+        
+        if let cacheKeyValue = cacheKeyValue {
+            _ = try cache.put(cacheKeyValue.value, in: cacheKeyValue.query).result.get()
+        }
+    }
+    
+    func test_getValue_withMain_withEmptyMain() throws {
+        // Given
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: Should be a random string

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+        let operation = MainOperation()
+        
+        // When
+        do {
+            _ = try cacheRepository.get(query, operation: operation).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.NotFound.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: Could be better to follow another style to check the error, see the in the `InMemoryDataSourceTests`.

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        expect(self.cacheSpy.getCalls.count).to(equal(0))
+        expect(self.mainSpy.getCalls.count).to(equal(1))
+        expect(self.mainSpy.getCalls[0]).to(be(query))
+    }
+    
+    func test_getValue_withMain_withMainValue() throws {
+        // Given
+        let value = Int.random()
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: should be random

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        // Then
+        expect(result).to(equal(value))
+        expect(self.cacheSpy.getCalls.count).to(equal(0))
+        expect(self.mainSpy.getCalls.count).to(equal(1))
+        expect(self.mainSpy.getCalls[0]).to(be(query))
+    }
+    
+    func test_getValue_withCache_withEmptyCache() throws {
+        // Given
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: should be random

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+        let operation = CacheOperation()
+        
+        // When
+        do {
+            _ = try cacheRepository.get(query, operation: operation).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.NotFound.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: same as above

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+        expect(self.mainSpy.getCalls.count).to(equal(0))
+        expect(self.cacheSpy.getCalls.count).to(equal(1))
+        expect(self.cacheSpy.getCalls[0]).to(be(query))
+    }
+    
+    func test_getValue_withCache_withCacheValue() throws {
+        // Given
+        let value = Int.random()
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: should be random

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        expect(result).to(equal(value))
+        expect(self.mainSpy.getCalls.count).to(equal(0))
+        expect(self.cacheSpy.getCalls.count).to(equal(1))
+        expect(self.cacheSpy.getCalls[0]).to(be(query))
+    }
+    
+    func test_getValue_withCacheSync_withEmptyCache_withMainValue() throws {
+        // Given
+        let value = Int.random()
+        let query = IdQuery("id")
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: should be random

File: Tests/HarmonyTests/CacheRepositoryTests.swift [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+        let operation = CacheSyncOperation()
+        
+        // When
+        do {
+            _ = try cacheRepository.get(query, operation: operation).result.get()
+            XCTAssert(false)
+        } catch {
+            // Then
+            expect(error).to(beAnInstanceOf(CoreError.NotFound.self))
+        }
```
Comment (joselufo) [inline Tests/HarmonyTests/CacheRepositoryTests.swift]: same as above

**Review by joselufo — CHANGES_REQUESTED** [review]

File: Tests/HarmonyTests/InMemoryDataSourceTests.swift (hunk 1 above)
Comment (vilanovi) [reply Tests/HarmonyTests/InMemoryDataSourceTests.swift]: fixed

**Review by joselufo — APPROVED** [review]

**Pull request merged**